### PR TITLE
[Merged by Bors] - feat: computation of `Over A` for a presheaf `A`

### PR DESCRIPTION
**Pull request by TwoFX**
---

Perhaps the worst De Bruijn factor I have ever seen, 10 lines of pen-and-paper proof turning into 575 lines of Lean code. I could probably split this file into multiple PRs, but there isn't really a point given that everything in the file except for the last two definition should be considered an implementation detail.

There is some relevant discussion here: https://github.com/leanprover-community/mathlib4/pull/10244#discussion_r1477365199

- [x] depends on: #10243 
- [x] depends on: #10244 

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/CategoryTheory/Comma/Presheaf.lean [hunk 1]
```diff
@@ -0,0 +1,575 @@
...
+                             ⇘           ⥥
+
+                               PSh(CostructuredArrow yoneda A)
+```
+
+where the top arrow is the forgetful functor forgetting the yoneda-costructure, the right arrow is
+the aforementioned equivalence and the diagonal arrow is the Yoneda embedding.
+
+In the notation of Kashiwara-Schapira, the type of the equivalence is written `C^ₐ ≌ Cₐ^`, where
+`·ₐ` is `CostructuredArrow` (with the functor `S` being either the identity or the Yonenda
```
Comment (joelriou) [inline Mathlib/CategoryTheory/Comma/Presheaf.lean:28]:
```suggestion
`·ₐ` is `CostructuredArrow` (with the functor `S` being either the identity or the Yoneda
```

File: Mathlib/CategoryTheory/Comma/Presheaf.lean [hunk 2]
```diff
@@ -0,0 +1,575 @@
...
+
+/-! ### Construction of the forward functor `Over A ⥤ (CostructuredArrow yoneda A)ᵒᵖ ⥤ Type v` -/
+
+/-- Via the Yoneda lemma, `u : F.obj (op X)` defines a natural transformation `yoneda.obj X ⟶ F`
+    and via the element `η.app (op X) u` also a morphism `yoneda.obj X ⟶ A`. This structure
+    witnesses the fact that these morphisms from a commutative triangle with `η : F ⟶ A`, i.e.,
+    that `yoneda.obj X ⟶ F` lifts to a morphism in `Over A`. -/
+structure MakesOverArrow {F : Cᵒᵖ ⥤ Type v} (η : F ⟶ A) {X : C} (s : yoneda.obj X ⟶ A)
+    (u : F.obj (op X)) : Prop where
+  (app : η.app (op X) u = yonedaEquiv s)
```
Comment (joelriou) [inline Mathlib/CategoryTheory/Comma/Presheaf.lean:85]: In mathlib4, it seems to me the usual syntax would be:
```suggestion
  app : η.app (op X) u = yonedaEquiv s
```

File: Mathlib/CategoryTheory/Comma/Presheaf.lean [hunk 3]
```diff
@@ -0,0 +1,575 @@
...
+def costructuredArrowIso (s t : CostructuredArrow yoneda A) : OverArrows s.hom t.hom ≅ t ⟶ s where
+  hom p := CostructuredArrow.homMk p.val (by aesop_cat)
+  inv f := yonedaArrow f.left f.w
+
+end OverArrows
+
+/-- This is basically just `yoneda.obj η : (Over A)ᵒᵖ ⥤ Type (max u v)` restricted along the
+    forgetful functor `CostructuredArrow yoneda A ⥤ Over A`, but done in a way that we land in a
+    smaller universe. -/
+@[simps (config := { fullyApplied := false }) obj map]
```
Comment (joelriou) [inline Mathlib/CategoryTheory/Comma/Presheaf.lean:200]: I am not sure we need this `fullyApplied := false`. The automatically generated simp lemmas look good to me (not just here, but throughout the file):
```suggestion
@[simps obj map]
```

Comment (TwoFX) [reply Mathlib/CategoryTheory/Comma/Presheaf.lean]: Thanks, I added these to work around some automation issues but it looks like it's no longer needed in the final version of the proof.

**Comment by joelriou** [comment]
Thanks!

bors merge

**Pull request closed without merging**